syntax = "proto2";

package newsfeed.proto;

enum topic_action_t {
    subscribe = 1;
    unsubscribe = 2;
};

enum global_error_t {
    ok = 1;
    not_registered = 2;
    internal = 3;
};

message register_request {
    required string userid = 1;
};

message register_response {
    required global_error_t error = 1;
    optional string topic = 2;
};

message topic_request {
    required topic_action_t action = 1;
    // if the topic does not exists then we will create one and subscribe
    required string topic = 2;
};

message topic_response {
    required topic_action_t action = 1;
    required global_error_t error = 2;
};

message post_news_request {
    required string news = 1;
};

message post_news_response {
    required global_error_t error = 1;
};

message news {
    required string data = 1;
};

message req_envelope {
    enum msg_type {
        register_request_t = 1;
        register_response_t = 2;
        topic_request_t = 3;
        topic_response_t = 4;
        post_news_request_t = 5;
        post_news_response_t = 6;
        news_t = 7;
    };

    required msg_type type = 1;
    optional register_request reg_req = 2;
    optional register_response reg_resp = 3;
    optional topic_request topic_req = 4;
    optional topic_response topic_resp = 5;
    optional post_news_request post_req = 6;
    optional post_news_response post_resp = 7;
    optional news news_data = 8;
};
